runtime/domain.c

```diff
@@ -374,8 +384,74 @@ struct domain_startup_params {
   uintnat unique_id;
 };
 
+static void* backup_thread_func(void* v)
+{
+  dom_internal* di = (dom_internal*)v;
+  uintnat msg;
+  struct interruptor* s = &di->interruptor;
+
+  domain_self = di;
+  SET_Caml_state((void*)(di->tls_area));
+
+  caml_plat_lock (&di->domain_lock);
+  while (1) { /* loop1 */
+    msg = atomic_load_acq (&di->backup_thread_msg);
+    Assert (msg <= BT_TERMINATE);
+    if (msg == BT_ENTERING_OCAML) {
+      /* Main thread wants to enter OCaml */
+      caml_plat_wait(&di->domain_cond);
+    } else if (msg == BT_IN_BLOCKING_SECTION) {
+      /* Handle interrupts on behalf of the main thread */
+      caml_plat_lock(&s->lock);
+      /* Both [s->lock] and [di->domain_lock] held here */
+      while (1) { /* loop2 */
+        msg = atomic_load_acq (&di->backup_thread_msg);
+        Assert (msg <= BT_TERMINATE);
+        if (msg == BT_ENTERING_OCAML) {
+          /* Main thread is leaving blocking section and entering OCaml */
+          caml_plat_unlock (&s->lock);
+          break; /* break loop2 and goto loop1 */
+        } else if (handle_incoming(s) == 0) {
+          caml_plat_wait(&s->cond);
+        }
+      }
+    } else if (msg == BT_TERMINATE) {
+      atomic_store_rel(&di->backup_thread_msg, BT_INIT);
+      caml_plat_unlock (&di->domain_lock);
+      break;
+    }
+  }
+  return 0;
+}
+
+static void install_backup_thread (dom_internal* di)
+{
+  int err;
+
+  if (di->backup_thread_running == 0) {
+    Assert (di->backup_thread_msg == BT_INIT ||     /* Using fresh domain */
+            di->backup_thread_msg == BT_TERMINATE); /* Reusing domain */
+
+    while (atomic_load_acq(&di->backup_thread_msg) != BT_INIT) {
+      /* Give a chance for backup thread on this domain to terminate */
+      caml_plat_unlock (&di->domain_lock);
```

Comment:
I couldn't work out if we need to add this:
` caml_plat_signal(&di->domain_cond);`
or if the pthread conditional signalling from `domain_terminate` will always wake the backup thread even if the `domain_lock` then ends up taken by this thread reusing the domain.

Reply:
The signal isn't necessary here. If the backup thread had been created, it is either waiting to obtain the lock, or if it has got the lock, then it is waiting on the condition variable (since the `domain_terminate` code holds the `domain_lock`). In the former case, eventually, the backup thread will get the lock, read the message, and go on to terminate, setting the `backup_thread_msg` to `BT_INIT`. In the latter case, it will wake up certainly and will terminate.

Reply:
Ok, that makes sense. My concern was more around `pthread_cond_wait` and if that could somehow break. That's actually fine because `pthread_cond_signal` will push the backup thread to be waiting on the mutex.